cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperatorTest.java

```diff
@@ -1451,6 +1451,60 @@ public void testCruiseControlDisabledToEnabledBehaviour(VertxTestContext context
                 }));
     }
 
+    @Test
+    public void testKrWhenCCUserTaskIsEmpty(VertxTestContext context) throws IOException, URISyntaxException {
+
+        cruiseControlServer.setupCCRebalanceResponse(0, CruiseControlEndpoints.REBALANCE);
+
+        KafkaRebalance kr = new KafkaRebalanceBuilder(createKafkaRebalance(namespace, CLUSTER_NAME, RESOURCE_NAME, EMPTY_KAFKA_REBALANCE_SPEC, false))
+                .withNewStatus()
+                .withObservedGeneration(1L)
+                .withConditions(new ConditionBuilder()
+                        .withType("ProposalReady")
+                        .withStatus("True")
+                        .build())
+                .endStatus()
+                .build();
+
+        Crds.kafkaRebalanceOperation(client).inNamespace(namespace).resource(kr).create();
+        Crds.kafkaRebalanceOperation(client).inNamespace(namespace).resource(kr).updateStatus();
```

Comment:
Why do we need this?

Reply:
It is there so we can directly move the Rebalance resource to `ProposalReady` state.

Reply:
But you are already creating the resource with `ProposalReady` state.

Reply:
Yes but to reflect that state we require the above two things if I recall correctly

Reply:
Not when using a proper Kubernetes API Server. I think Shubham has this right.

Reply:
I asked because I wasn't sure and it seems to work without it.

Reply:
So understanding it in a bit more depth I see that the logs change in both the cases. If you remove 
`  Crds.kafkaRebalanceOperation(client).inNamespace(namespace).resource(kr).updateStatus();`Looking into it more, I guess the proposal never moved to `PendingProposal` state and directly moved to Proposal ready?

Reply:
I guess there is a difference, If you remove         `Crds.kafkaRebalanceOperation(client).inNamespace(namespace).resource(kr).updateStatus();` , the logs says 
```
2024-06-24 21:45:15 INFO  AbstractOperator:264 - Reconciliation #1(test-trigger) KafkaRebalance(testkrwhenccusertaskisempty/my-rebalance): KafkaRebalance my-rebalance will be checked for creation or modification
2024-06-24 21:45:15 INFO  KafkaRebalanceAssemblyOperator:370 - Reconciliation #1(test-trigger) KafkaRebalance(testkrwhenccusertaskisempty/my-rebalance): Rebalance action is performed and KafkaRebalance resource is currently in [New] state
2024-06-24 21:45:15 INFO  KafkaRebalanceAssemblyOperator:1170 - Reconciliation #1(test-trigger) KafkaRebalance(testkrwhenccusertaskisempty/my-rebalance): Requesting Cruise Control rebalance [dryrun=true]
2024-06-24 21:45:16 INFO  CrdOperator:123 - Reconciliation #1(test-trigger) KafkaRebalance(testkrwhenccusertaskisempty/my-rebalance): Status of KafkaRebalance my-rebalance in namespace testkrwhenccusertaskisempty has been updated
2024-06-24 21:45:16 INFO  AbstractOperator:536 - Reconciliation #1(test-trigger) KafkaRebalance(testkrwhenccusertaskisempty/my-rebalance): reconciled
2024-06-24 21:45:16 INFO  AbstractOperator:264 - Reconciliation #2(test-trigger) KafkaRebalance(testkrwhenccusertaskisempty/my-rebalance): KafkaRebalance my-rebalance will be checked for creation or modification
2024-06-24 21:45:16 INFO  KafkaRebalanceAssemblyOperator:370 - Reconciliation #2(test-trigger) KafkaRebalance(testkrwhenccusertaskisempty/my-rebalance): Rebalance action is performed and KafkaRebalance resource is currently in [ProposalReady] state
```
But if you don't remove that part the logs are different
```
2024-06-24 21:46:32 INFO  AbstractOperator:264 - Reconciliation #1(test-trigger) KafkaRebalance(testkrwhenccusertaskisempty/my-rebalance): KafkaRebalance my-rebalance will be checked for creation or modification
2024-06-24 21:46:32 INFO  KafkaRebalanceAssemblyOperator:370 - Reconciliation #1(test-trigger) KafkaRebalance(testkrwhenccusertaskisempty/my-rebalance): Rebalance action is performed and KafkaRebalance resource is currently in [ProposalReady] state
2024-06-24 21:46:32 INFO  AbstractOperator:536 - Reconciliation #1(test-trigger) KafkaRebalance(testkrwhenccusertaskisempty/my-rebalance): reconciled
2024-06-24 21:46:32 INFO  AbstractOperator:264 - Reconciliation #2(test-trigger) KafkaRebalance(testkrwhenccusertaskisempty/my-rebalance): KafkaRebalance my-rebalance will be checked for creation or modification
```

Reply:
So basically in the first case the rebalance resource never moved to `PendingProposal` and directly to `ProposalReady` from `New` state?

Reply:
@fvaleri Just to clarify ... In Kubernetes, a regular POST/PUT/PATH request does not change the `.status` section. You have a special request for it. So if you do only `create()`, you will have only the `.metadata` and `.spec` set. You need to do also the `updateStatus()` to set the `.status` part. So from Kube perspective, if you want to prepare the custom resource with some specific status, you have to do both calls. Whether you really need the status and what role it plays is up to the test of course.  So in general, these two calls might make sense, but I personally didn't study the test to see how it uses it.

(Also, one needs to be careful about race conditions here. It is easy to do this in preparation. But if the test is already running, it creates two separate events.)

Reply:
@scholzj that's also my understanding. The point I was trying to raise is that I don't think we need to set the status when creating the KR as we then call reconcile. Translating this into code, I think we should simply do:

```sh
KafkaRebalance kr = createKafkaRebalance(namespace, CLUSTER_NAME, RESOURCE_NAME, EMPTY_KAFKA_REBALANCE_SPEC, false);
Crds.kafkaRebalanceOperation(client).inNamespace(namespace).resource(kr).create();
```

Wdyt?

Reply:
In general, you might have two reasons to do something like this:
* Improve the speed of the tests by mocking the initial state where the test starts (for example here it starts with a ready proposal, so I guess it avoids at least two reconciliations to get there?).
* Reduce the amount of required mocking. By starting already in some particular state, you do not need to mock how to get the proposal.

So I think it is valid in general. But if the test passes with it commented out, it seems a bit strange to me and suggests some flawed logic.

Reply:
I see the proposal moving directly from "New" - "ProposalReady" at the first reconciliation called by the test. I guess this is the effect of how the mock response has been setup, so it should be fine. That's how you improve the speed of the test in this case.

Reply:
I have configured the Mock response to return the `PendingProposal` first instead of directly moving to `ProposalReady`